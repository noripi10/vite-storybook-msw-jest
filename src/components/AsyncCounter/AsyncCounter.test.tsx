import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
// import userEvent from '@testing-library/user-event';

import { AsyncCounter } from './';

// const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('AsyncCounter', () => {
  beforeEach(() => {
    const { asFragment } = render(<AsyncCounter />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('初期化', () => {
    const count = screen.getByTestId('async_count');
    expect(count.innerHTML).toBe('0');
  });

  test('インクリメント', async () => {
    jest.useFakeTimers();
    const button = screen.getByText('count up');
    fireEvent.click(button);
    act(() => {
      jest.runAllTimers();
    });
    const count = screen.getByTestId('async_count');
    expect(count.innerHTML).toBe('1');
    jest.useRealTimers();
  });

  test('デクリメント', async () => {
    jest.useFakeTimers();
    const button = screen.getByText('count down');
    fireEvent.click(button);
    act(() => {
      jest.runAllTimers();
    });
    const count = screen.getByTestId('async_count');
    expect(count.innerHTML).toBe('-1');
    jest.useRealTimers();
  });

  test('押下時のボタン非活性', () => {
    const button = screen.getByRole('button', { name: 'count up' });
    fireEvent.click(button);
    expect(button).toBeDisabled();
  });

  test('押下後のボタン活性', () => {
    jest.useFakeTimers();
    const button = screen.getByRole('button', { name: 'count up' });
    fireEvent.click(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(button).toBeEnabled();
    jest.useRealTimers();
  });

  test('押下時にロードが表示', () => {
    const button = screen.getByRole('button', { name: 'count up' });
    fireEvent.click(button);
    expect(screen.getByText('loading...')).toBeEnabled();
  });

  test('押下後にロードが非表示', () => {
    jest.useFakeTimers();
    const button = screen.getByRole('button', { name: 'count up' });
    fireEvent.click(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(screen.queryByText('loading...')).not.toBeInTheDocument();
    jest.useRealTimers();
  });
});
