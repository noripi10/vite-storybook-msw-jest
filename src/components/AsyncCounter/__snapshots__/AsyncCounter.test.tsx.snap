// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncCounter インクリメント 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter デクリメント 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter 初期化 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter 押下後にロードが非表示 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter 押下後のボタン活性 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter 押下時にロードが表示 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;

exports[`AsyncCounter 押下時のボタン非活性 1`] = `
<DocumentFragment>
  <div>
    <strong
      data-testid="async_count"
    >
      0
    </strong>
  </div>
  <button>
    count down
  </button>
  <button>
    count up
  </button>
</DocumentFragment>
`;
